### DISCUSSION TITLE
What triggers `workspace.onDidChangeTextDocument`?

### DISCUSSION THREAD

**Discussion by worksofliam** [microsoft/vscode-discussions#90] [Extension Development QnA] [answered]
Hi!

I am wondering what triggers `workspace.onDidChangeTextDocument`? Of course, a user saving a file would trigger it, but what if I use git to discard changes against a file (or many files) - will that also trigger it?

Thanks!

**Comment by alefragnani** [comment, marked as answer]
Hi @worksofliam ,

Based on its documentation (https://code.visualstudio.com/api/references/vscode-api) it says: 

> An event that is emitted when a [text document](https://code.visualstudio.com/api/references/vscode-api#TextDocument) is changed. This usually happens when the [contents](https://code.visualstudio.com/api/references/vscode-api#TextDocument.getText) changes but also when other things like the [dirty](https://code.visualstudio.com/api/references/vscode-api#TextDocument.isDirty)-state changes.

I would expand this description with a few scenarios: It is triggered when:

* When the user types something
* When `Undo` and `Redo` are fired
* When `Save` is fired
* When Formatters are fired

And **yes**, also when _discarding changes with Git_

Hope this helps

**Reply by worksofliam** [reply]
Thanks! It does. I appreciate it.

**Reply by alefragnani** [reply]
That's great to hear!

Please, if you don't mind, mark the answer as _accepted_, so other users looking for the similar questions will know there is an answer here.

Thank you

**Comment by r3m0t** [comment]
TextDocument.contents is "what should the text editor show for this file". 

VS Code watches files and updates its view of their contents even when an external program runs like `git restore`, but if the file is dirty (user has unsaved edits in VSCode), this doesn't happen immediately because it only has one TextDocument.contents and the user's input is the most important version. (A notification shows up on save instead).

Also, if the document doesn't have an editor open, there might not be a TextDocument object for that filename.
